pkg/infra/observer.go

```diff
@@ -5,62 +5,55 @@ import (
 	"time"
 
 	"github.com/hyperledger/fabric-protos-go/peer"
+	"github.com/pkg/errors"
 	log "github.com/sirupsen/logrus"
 )
 
 type Observer struct {
 	d      peer.Deliver_DeliverFilteredClient
 	logger *log.Logger
-	signal chan error
 }
 
-func CreateObserver(channel string, node Node, crypto *Crypto, logger *log.Logger) *Observer {
-	deliverer, err := CreateDeliverFilteredClient(node)
+func CreateObserver(channel string, node Node, crypto *Crypto, logger *log.Logger) (*Observer, error) {
+	deliverer, err := CreateDeliverFilteredClient(node, logger)
 	if err != nil {
-		panic(err)
+		return nil, err
 	}
 
 	seek, err := CreateSignedDeliverNewestEnv(channel, crypto)
 	if err != nil {
-		panic(err)
+		return nil, err
 	}
 
 	if err = deliverer.Send(seek); err != nil {
-		panic(err)
+		return nil, err
 	}
 
 	// drain first response
 	if _, err = deliverer.Recv(); err != nil {
-		panic(err)
+		return nil, err
 	}
 
-	return &Observer{d: deliverer, signal: make(chan error, 10), logger: logger}
+	return &Observer{d: deliverer, logger: logger}, nil
 }
 
-func (o *Observer) Start(N int, now time.Time) {
-	defer close(o.signal)
+func (o *Observer) Start(N int, errorC chan error, finishCh chan int, now time.Time) {
 	o.logger.Debugf("start observer")
 	n := 0
 	for n < N {
 		r, err := o.d.Recv()
 		if err != nil {
-			o.signal <- err
+			errorC <- err
 		}
 
 		if r == nil {
-			panic("Received nil message, but expect a valid block instead. You could look into your peer logs for more info")
+			errorC <- errors.Errorf("received nil message, but expect a valid block instead. You could look into your peer logs for more info\n")
+			return
 		}
 
 		fb := r.Type.(*peer.DeliverResponse_FilteredBlock)
 		n = n + len(fb.FilteredBlock.FilteredTransactions)
 		fmt.Printf("Time %8.2fs\tBlock %6d\tTx %6d\n", time.Since(now).Seconds(), fb.FilteredBlock.Number, len(fb.FilteredBlock.FilteredTransactions))
 	}
-}
-
-func (o *Observer) Wait() {
-	for err := range o.signal {
-		if err != nil {
-			o.logger.Errorf("Observed error: %s\n", err)
-		}
-	}
+	finishCh <- 1
```

Comment:
why don't we simply close it?

Reply:
I don't want to close the channel here.
Considering with multiple observers with different peers, the one close in the sub goroutine will make other goroutine close a closed channel, I would like with `<- 1` or `<- struct{}{}` ?
as so far either `<-` or `close` are the same.

Reply:
`finishCh` should **not** be shared - one `observer` should **only** cares about its own target and close channel when completed. It's really up to the user of `observer` to decide what to do about it. In the case you described, when we need to monitor **multiple** `observers, you can simply start several goroutines and block with `WaitGroup`

Reply:
k